RSC can create hospitals

angiecstn opened the pull request:
### PT Link
https://www.pivotaltracker.com/story/show/161589429

### Included in PR

* Adding create and show method for Hospital
* Fixing routes for Hospital
* Fixing view for Hospital
* Unit test + Feature test

### Screenshot

<img width="1440" alt="screenshot 2018-11-05 at 20 44 55" src="https://user-images.githubusercontent.com/41902068/48023144-3b5ceb00-e13d-11e8-8463-2ae63aa3cd94.png">

angiecstn committed:
adding create and show actions for hospitals

angiecstn committed:
adding features for rsc can add hospitals

Noelryn committed:
Added in steps for rsc can create hospital

angiecstn committed:
Merge branch 'development' of https://github.com/CraftAcademy/nurse_and_doc into rsc_create_hospitals

angiecstn committed:
adding routes for hospital show, new and create

tochman commented on features/rsc_can_create_new_hospitals.feature:
```diff
@@ -0,0 +1,29 @@
+    Scenario: RSC cannot add hospitals if it already exists
+        Given I am on the landing page
+        And I click on "Hospitals"
+        Then I see "Hospital 1"
```
`Then I see`? not `Then I should see` ??

angiecstn replied:
We fixed it

angiecstn committed:
fixing routes and refactoring steps in scenario

angiecstn committed:
adding missin route

angiecstn committed:
adding link to hospitals view

rupertlion committed:
Merge branch 'rsc_create_hospitals' of https://github.com/angiecstn/nurse_and_doc into rsc_create_hospitals

rupertlion committed:
Trying to get params to send through link_to helper

Noelryn committed:
Changed Hospitals to @hospitals on index partial

Noelryn committed:
Added in forms to index partials and javascript into index js tests still failing but looking better

Noelryn committed:
Added in the hospital path to the index pprtial for hospitals and also refactored the test for the the steps to find hospitals on the hospitals and add the hospital path to the add button with no effect unfortunely

Noelryn committed:
Refactored the from and have it working currently not saving but can click on the hospital page and see the form

Noelryn committed:
Added in a show method to the controller but there is still no save method

angiecstn committed:
modifying controller to save hospitals and render new hospitals in view

tochman commented on app/views/hospitals/_new.html.haml:
```diff
@@ -0,0 +1,17 @@
+        #form-button
+        =link_to "Add", "#", class: "button"
+        = button_tag 'Cancel', id: 'cancel_modal', class: 'signup-button'
+        
```
Don't forget to render the partial

Noelryn committed:
Added ds store into git ignore and changed the sad path in the feature steps one test falling

Noelryn committed:
Added the javascript pop up to the new html partial and the to the new js.erb, also refactroed the controller and update the hospital index link

angiecstn committed:
modifying controller, creating new.js and _new.html.haml, modifying feature steps to add ajax flow

Noelryn committed:
Pull down from angie and had some merge conflicts fixed error messages

Noelryn committed:
Removed instance variables from line 11 and 12 in the controller and added a new step to for the success message

RhonGabriel committed:
added uniqueness of hospital namel, changed hospital view

RhonGabriel committed:
modified hospital model to include uniqueness of names

angiecstn committed:
fixing merge conflicts

angiecstn committed:
Merge branch 'rsc_create_hospitals' of https://github.com/angiecstn/nurse_and_doc into rsc_create_hospitals

angiecstn committed:
fixing view, removing files not needed, fixing message in feature

angiecstn committed:
fixing merge conflicts, adding validation for name uniqueness

RhonGabriel commented on app/views/hospitals/_new.html.haml:
```diff
@@ -0,0 +1,13 @@
+-#                 = form.submit value: "Submit", class: "bg-blue hover:bg-blue-dark text-white font-bold py-2 px-4 my-4 rounded focus:outline-none focus:shadow-outline"
+-#                 = button_tag 'Cancel', id: 'cancel_modal', class: 'signup-button'
+
+-# = render partial: 'partials/modal'
```
Since you are not using the modal anymore, should you cancel this file?

RhonGabriel commented on features/step_definitions/assertion_steps.rb:
```diff
@@ -34,6 +34,9 @@
 
+Then("I see {string}") do |content|
+    expect(page).to have_content content
+  end
```
indentation to be fixed

RhonGabriel committed:
run db migrate

angiecstn committed:
refactoring

Prozak8 commented on app/controllers/hospitals_controller.rb:
```diff
@@ -2,4 +2,30 @@ class HospitalsController < ApplicationController
+
+  def new
+    @hospital = Hospital.new
+  end
```
The website works without this

Prozak8 commented on app/controllers/hospitals_controller.rb:
```diff
@@ -2,4 +2,30 @@ class HospitalsController < ApplicationController
+
+  def show
+    @hospitals = Hospital.find(params[:id])
+  end
```
You don't have a show page. No need for this.

Noelryn replied:
I have removed  the show action

Prozak8 commented on config/routes.rb:
```diff
@@ -9,6 +9,6 @@
     resources :users, only: [:create, :new]
     resources :staffcos, only: :index
-    resources :hospitals, only: :index
+    resources :hospitals, only: [:index, :create, :new]
```
If you take away :new it will still work

Noelryn replied:
This has also been removed

Prozak8 commented on features/rsc_can_create_new_hospitals.feature:
```diff
@@ -0,0 +1,30 @@
+    Background: 
+        Given the following hospitals exists
+            | name        | 
+            | Hospital 1  | 
```
You only need hospital 2

Prozak8 replied:
You could also include the first 3 lines of each scenario in here

Noelryn committed:
Removed the show and new actions from the hospitals controller,
Removed the new route for the rails routes,
Finally removed the Hospital 1 name from the background steps in the rsc can create a new hospital feature tests

Noelryn committed:
Added the new route back into the routes

angiecstn committed:
fixing merge conflicts

angiecstn committed:
refactoring

Noelryn committed:
Refactored the feature test changed add hospitals to name and refactored the index partial for hospitals all working

Noelryn committed:
Refactoring feature tests, moving tests from the sad path into the background steps

tochman commented on app/views/hospitals/_index.html.haml:
```diff
@@ -6,7 +6,11 @@
+    #list-grid
+      #list-name
+      = f.text_field :name, id: "Name", class: "form-input-field"
+      #list-button-area
```
I still don't understand why this is an id and not a class. (Same with `list-grid` and `list-name`)

tochman commented on features/rsc_can_create_new_hospitals.feature:
```diff
@@ -0,0 +1,29 @@
+            | Hospital 2  |
+        Given I am on the landing page
+        And I click on "Hospitals"
+        Then I see "Hospital"
```
There's no need for this assertion. We discussed this before. There is no need to assertions in backgrounds

tochman commented on features/rsc_can_create_new_hospitals.feature:
```diff
@@ -0,0 +1,29 @@
+    Scenario: RSC can successfully add new hospitals
+        Given I am on the landing page
+        And I click on "Hospitals"
+        Then I should see "Hospital"
```
You are repeating steps that are already in the background. Also, see my comment about the assertion.

tochman replied:
I still see this assertion, why?

tochman commented on features/step_definitions/assertion_steps.rb:
```diff
@@ -34,6 +34,10 @@
     expect(page).to have_current_path(root_path)
 end
 
+Then("I see {string}") do |content|
```
Again, an assertion should contain the word `should`  ;-)

tochman commented on features/step_definitions/basic_steps.rb:
```diff
@@ -28,6 +28,10 @@
   }
 end
 
+Then("I fill in {string}") do |content|
```
`When`, please use `When` in these kind of steps

tochman commented on features/step_definitions/basic_steps.rb:
```diff
@@ -28,6 +28,10 @@
 end
 
+Then("I fill in {string}") do |content|
+  fill_in with: content 
```
`fill_in` what? What field are you filling in? The way this step is defined you might get a false positive.

angiecstn replied:
We have now removed it since it was unnecessary and we fixed the test accordingly.

tochman commented on spec/models/hospital_spec.rb:
```diff
@@ -4,7 +4,7 @@
 
   describe 'Validations' do
-    it { is_expected.to validate_presence_of :name }
+    it { should validate_uniqueness_of(:name) }
```
Please do NOT mix in old should` syntax in rspec. Do more research. You can use the `validate_uniqueness_of` matcher with the `is_expected.to` syntax.

RhonGabriel committed:
pull from upstream, fix merge conflict

RhonGabriel committed:
changed tailwind id to classes

RhonGabriel committed:
removed assertions from background in feature test

angiecstn committed:
fixing spec file

RhonGabriel committed:
fixed fail on feature test not finding field name

tochman reviewed (dismissed):
Please address my comments before removing the WIP tag and merge. Not quite ready for merge yet IMO.

angiecstn committed:
fixing feature steps for job creation

angiecstn committed:
removing unneccessary basic step

angiecstn committed:
removing unneeded steps, refactoring

angiecstn committed:
Merge branch 'development' of https://github.com/CraftAcademy/nurse_and_doc into rsc_create_hospitals

angiecstn committed:
fixing merge conflicts

angiecstn committed:
removing white space

tochman commented on app/views/hospitals/_index.html.haml:
```diff
@@ -6,7 +6,11 @@
+  = form_with url: hospitals_path, scope: :hospital  do |f|
+    .list-grid
+      .list-name
+      = f.text_field :name, id: "Name", class: "form-input-field"
```
If `.list-grid` is supposed to be a wrapper around the `link_to`, this line needs to be intended

angiecstn replied:
I cannot see a ``` link_to ``` in this file, I think the file changed when we merged the fix PR

tochman replied:
Sorry, my bad, I mean `f.text_field`

tochman replied:
```
.list-grid
      .list-name
             = f.text_field :name, id: "Name", class: "form-input-field"
```

tochman replied:
And also, please make sure to use single quotes. Install RuboCop extension and you will be able to format your code and fix details like that with auto format.

tochman commented on app/views/hospitals/_index.html.haml:
```diff
@@ -6,7 +6,11 @@
+      .list-name
+      = f.text_field :name, id: "Name", class: "form-input-field"
+      .list-button-area
+      = f.submit value: "Add Hospital", class: "button text-center"
```
See comment above

angiecstn committed:
adjustying indentation in new.html.haml

angiecstn committed:
removing unnecessary assertions in hospital feature

tochman commented on app/controllers/hospitals_controller.rb:
```diff
@@ -2,4 +2,22 @@ class HospitalsController < ApplicationController
   end
+
+  def create
+    unless Hospital.exists?(hospital_params)
```
Oh, I just spotted this. What is this doing for you? Isn't the `else` block handling the error?

angiecstn committed:
refactoring

tochman reviewed (approved):
Good job.
